nodewatcher/modules/monitor/http/survey/processors.py

```diff
@@ -0,0 +1,106 @@
+        # (master branch at https://github.com/wlanslovenija/nodewatcher-agent/blob/master/modules/wireless.c#L437).
+        # So a survey is performed once every two hours, meaning that we use hourly granularity.
+        return datastream.Granularity.Hours
+
+
+class SurveyInfoStreamsData(object):
+    def __init__(self, node, neighbor_graph):
+        self.node = node
+        self.neighbor_graph = neighbor_graph
+
+
+ds_pool.register(SurveyInfoStreamsData, SurveyInfoStreams)
+
+
+class SurveyInfo(monitor_processors.NodeProcessor):
+    """
+    Accesses the data collected during a WiFi survey of neighboring wireless access points / routers. Stores channel,
+    BSSID, SSID and signal strength for each neighbor into the datastream. A star shaped graph is constructed. The
+    source vertex contains an array of all BSSIDs of that node. Source vertex is stored according to its UUID,
+    wherease all neighbors are stored according to their BSSIDs. Will only run if HTTP monitor module has previously
+    fetched data.
+    """
+
+    @monitor_processors.depends_on_context("http", http_processors.HTTPTelemetryContext)
+    def process(self, context, node):
+        """
+        Called for every processed node.
+
+        :param context: Current context
+        :param node: Node that is being processed
+        :return: A (possibly) modified context
+        """
+
+        version = context.http.get_module_version('core.wireless')
+        if version == 0:
+            # Unsupported version or data fetch failed
+            return context
+
+        edges, vertices = [], []
+        try:
+            source_vertex_bssids = [interface.bssid for interface in context.http.core.wireless.interfaces.values()]
+            source_vertex = {
+                'i': str(node.uuid),
+                'bssids': source_vertex_bssids,
+            }
+            vertices.append(source_vertex)
+            for radio in context.http.core.wireless.radios.values():
+                for neighbor in radio.survey:
+                    vertices.append({'i': neighbor['bssid']})
+                    # Add an edge for that neighbor.
+                    edge = {'f': str(node.uuid), 't': neighbor['bssid']}
+                    for field in ('channel', 'signal', 'ssid'):
+                        if field in neighbor:
+                            edge[field] = neighbor[field]
+                    edges.append(edge)
+        except KeyError:
+            pass
+        except NameError:
+            self.logger.warning("Could not parse survey data for node: " + str(node))
+
+        latest_stored_graph = None
+        # Retrieve the latest stored datapoint, if it exists
+        streams = datastream.find_streams({'node': node.uuid})
+        if streams:
+            latest_stream = streams[len(streams) - 1]
+            datapoint_iterator = datastream.get_data(stream_id=latest_stream['stream_id'],
```

Comment:
> Also, by using reverse=True I assume that I would use an earlier start parameter and fetch more results and then choose the first one, correct? But why would we fetch more results than necessary (since we're using the default option of start_exclusive=None)?

No, you use the same `start`, you just use `reverse` so that you know that element at `[0]` is the latest one.

> But why would we fetch more results than necessary (since we're using the default option of start_exclusive=None)?

Also keep in mind that you are not fetching any results until you read them. You just prepare a cursor. You read them when you iterate over them. So `[0]` reads only the first result. (BTW, `len` might be expensive because it might force all elements to be read. I do not remember if we optimized it in this particular case, but in general for iterators `len` might be expensive.)